### DISCUSSION TITLE
VS Code doesn't recognise fenced code blocks that use attributes or when language is prefixed with a dot

### DISCUSSION THREAD

**Discussion by reenberg** [mkdocs/mkdocs#3521] [Q&A] [answered]
This is a VS Code related question in relation to writing my documentation using mkdocs-material.  So I know it is a bit off topic, but I figured someone here might have experienced this as well.

**Question**
Is there a way to get VS Code to recognise the MkDocs specifics fenced code blocks, and perform code highlighting in the editor?  
Does people just accept this or am I missing some extension to make it play nice with the MkDocs format?

**Background**
According to the [MkDocs documentation](https://github.com/Python-Markdown/markdown/blob/master/docs/extensions/fenced_code_blocks.md), on fenced code blocks, language specification should be prefixed with a dot:

> The language of the code within a code block can be specified for use by syntax highlighters, etc. The language should be prefixed with a dot and not contain any whitespace (.language-name).

but it may be excluded.  However when additional classes is desired, then the curly braces is mandatory, which also makes the dot prefix of the language mandatory.

> Additionally, the curly braces and dot are required for all classes, including the language class if more than one class is defined.

This is all great.  However code highlighting of the fenced code block stops working in VS Code once the language name is dot prefixed, and when the curly braced attribute list is used.

The following snippet gets highlighted

``````
``` bash
# Some comment
echo "Hello World"
```
``````

However this does not:

``````
``` .bash
# Some comment
echo "Hello World"
```
``````

and this neither

``````
``` { .bash } 
# Some comment
echo "Hello World"
```
``````

The only documentation I was able to find, was on the VS Code Markdown Preview, which says it is using `markdown-it` and Highlight.js.  
However it seems that the editor uses some other method to parse the Markdown, as I have found some examples of the editor not even syntax highlighting where the preview does (see below).

I know that the Preview is not really to be used/trusted, and I am using the `mkdocs serve` method to look at my rendered output.  I just mentioned it to make the point.  However the Preview can be nice sometimes for small stuff as it is readily available next to the editor.

``````
*   This will only highlight the "Hello World" string in the Editor, but will
    highlight correctly in the Preview:

    ``` bash
    # Some comment
    echo "Hello World"
    ```

*   This will mostly highlight in the Editor, except for the `echo`, but will
    highlight correctly in the Preview:

     ``` bash
     # Some comment
     echo "Hello World"
     ```

    Note that the above code block is indented one space

This highlights correctly in both the Editor and Preview:

``` bash
# Some comment
echo "Hello World"
```
``````

**Comment by oprypin** [comment]
Regarding the last part: I have no idea why it highlights correctly with 0 or 5 spaces but not with 4... That's clearly just a bug in VS Code highlighter. And I do observe your finding as well.

**Comment by oprypin** [comment, marked as answer]
My recommendation would be to just follow the widely liked format where the language always directly follows the backticks.

You must use [pymdownx.superfences](https://facelessuser.github.io/pymdown-extensions/extensions/superfences/)

This example is highlighted in VSCode, GitHub, and MkDocs+superfences (+it has a title above it in that case):

   ``````
   ```bash title="hello_world.sh"
   # Some comment
   echo "Hello World"
   ```
   ``````

```bash title="hello_world.sh"
# Some comment
echo "Hello World"
```

Though there *are* cases where this feature breaks, so I was so surprised  that it was failing for me just when I was sure that it always works.

Works:

```yaml
markdown_extensions:
  - pymdownx.superfences
```

Breaks:

```yaml
markdown_extensions:
  - pymdownx.superfences
  - pymdownx.highlight:
      use_pygments: false
```

@facelessuser is that expected btw?

**Reply by facelessuser** [reply]
What breaks, `title`? The `title` option leverages the Pygments option that creates special titles, so if you are not using Pygments, then anything you specify will be attached as attribute if you are using the `attr_list`, brace form. Your JS highlighter is responsible for adding any titles, line numbers, etc. according to its documentation and its configuration.

So, let's make this clear, assuming `attr_list` is enabled.

1. Using Pygments
    1. Each code block can specify special options that control the Pygments highlighter, such as `title` etc.
    2. Brace form allows attaching `#id` and `.classes`, all other attributes are processed as Pygments options unless they start with `data-`.
2. Not using Pygments
    1. Each code blocks accepts the language identifier.
    2. To add any `#id`, `.classes`, or `attr=value`, you must use the brace form.

This is all covered in the documentation: https://facelessuser.github.io/pymdown-extensions/extensions/superfences/#injecting-classes-ids-and-attributes.

````py
import markdown

# Option 1: Don't indent content of your string which triggers indented code blocks.
MD1 = """
```python title="title"
print('hello world')
```
"""

MD2 = """
```{.python title="title"}
print('hello world')
```
"""

html = markdown.markdown(
    MD1,
    extensions=['pymdownx.superfences', 'pymdownx.highlight', 'attr_list'],
    extension_configs={'pymdownx.highlight': {'use_pygments': True}}
)

print('=== No Brace Pygments ===')
print(html)


html = markdown.markdown(
    MD2,
    extensions=['pymdownx.superfences', 'pymdownx.highlight', 'attr_list'],
    extension_configs={'pymdownx.highlight': {'use_pygments': True}}
)

print('=== Brace Pygments ===')
print(html)

html = markdown.markdown(
    MD1,
    extensions=['pymdownx.superfences', 'pymdownx.highlight', 'attr_list'],
    extension_configs={'pymdownx.highlight': {'use_pygments': False}}
)

print('=== No Brace No Pygments ===')
print(html)

html = markdown.markdown(
    MD2,
    extensions=['pymdownx.superfences', 'pymdownx.highlight', 'attr_list'],
    extension_configs={'pymdownx.highlight': {'use_pygments': False}}
)

print('=== Brace No Pygments ===')
print(html)
````

```html
=== No Brace Pygments ===
<div class="highlight"><span class="filename">title</span><pre><span></span><code><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;hello world&#39;</span><span class="p">)</span>
</code></pre></div>
=== Brace Pygments ===
<div class="highlight"><span class="filename">title</span><pre><span></span><code><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;hello world&#39;</span><span class="p">)</span>
</code></pre></div>
=== No Brace No Pygments ===
<p><code>python title="title"
print('hello world')</code></p>
=== Brace No Pygments ===
<pre class="highlight"><code class="language-python" title="title">print('hello world')</code></pre>
```

**Reply by reenberg** [reply]
I managed to dig into the VS Code repository and follow that the build-in extension for Markdown has its TexMate grammar generated from: [microsoft/vscode-markdown-tm-grammar](https://github.com/microsoft/vscode-markdown-tm-grammar).  And I spottet this fairly new extension [libukai/mkdocs-highlight](https://github.com/libukai/mkdocs-highlight) which seems to try and add additional syntax for various MkDocs plugins.  One of those places could be a viable place to see if  support could be added.

> My recommendation would be to just follow the widely liked format where the language always directly follows the backticks.

Thanks @oprypin, I just really wanted the option to also add additional classes if needed, and hoped for some alternative that I had missed.  But from looking at the MS Markdown grammar linked above it seems that it is totally unsupported as of now.

> This example is highlighted in VSCode, GitHub, and MkDocs+superfences (+it has a title above it in that case):

I'm actually a bit surprised that this works without the curly braces of an attribute list, but then again the MkDocs documentation only states that the attribute list **should** be wrapped in curly braces, not that it must.  Only when using additional classes, the documentation states that it is **required**.

Jez, I'm constantly glad that I am not in the business of trying to parse Markdown in any structured way :)  Just looking at the open issues in the MS grammar repository :/

**Reply by reenberg** [reply]
Hm, they actually somewhat supports this.

[this commit ](https://github.com/microsoft/vscode-markdown-tm-grammar/commit/e073142da02667f4ab84bfdaea7570d14f82d377#diff-7d111b05fe05c50beb729d1e99a36010fbda8c89b57a1a26ac3e7b0778a0ed53L55) adds support for Codebraid annotations for some of the languages.  However the grammar is kind of horrible, as you are not allowed to have a space after the opening curley, and you need at least a space before the closing if you don't have any other attributes...


``````
``` {.bash }
# This works
```
``````

But again, it only works for some of the languages that it currently supports.  I assume the languages that Codebraid support. 

Oh well.  Just wanted to add that as a finishing remark to my dive into VS Code syntax highlighting.
